demos/multi_material/thermochemical_buoyancy/van_keken_1997.py

```diff
@@ -0,0 +1,277 @@
+)
+
+# This problem has a constant pressure nullspace, handled identically to our previous
+# tutorials.
+
+Z_nullspace = create_stokes_nullspace(Z)
+
+# Boundary conditions are specified next: free slip on all sides, heating from below,
+# and cooling from above.
+
+stokes_bcs = {
+    bottom_id: {"uy": 0},
+    top_id: {"uy": 0},
+    left_id: {"ux": 0},
+    right_id: {"ux": 0},
+}
+temp_bcs = {bottom_id: {"T": 1}, top_id: {"T": 0}}
+
+# We move on to initialising the temperature field.
+
+# +
+X = SpatialCoordinate(mesh)  # Extract UFL representation of spatial coordinates
+
+# Calculate quantities linked to the temperature initial condition using UFL
+u0 = lx ** (7 / 3) / (1 + lx**4) ** (2 / 3) * (Ra / 2 / sqrt(pi)) ** (2 / 3)
+v0 = u0
+Q_ic = 2 * sqrt(lx / pi / u0)
+Tu = erf((1 - X[1]) / 2 * sqrt(u0 / X[0])) / 2
+Tl = 1 - 1 / 2 * erf(X[1] / 2 * sqrt(u0 / (lx - X[0])))
+Tr = 1 / 2 + Q_ic / 2 / sqrt(pi) * sqrt(v0 / (X[1] + 1)) * exp(
+    -(X[0] ** 2) * v0 / (4 * X[1] + 4)
+)
+Ts = 1 / 2 - Q_ic / 2 / sqrt(pi) * sqrt(v0 / (2 - X[1])) * exp(
+    -((lx - X[0]) ** 2) * v0 / (8 - 4 * X[1])
+)
+
+# Interpolate temperature initial condition and ensure boundary condition values
+T.interpolate(max_value(min_value(Tu + Tl + Tr + Ts - 3 / 2, 1), 0))
+DirichletBC(Q, 1, bottom_id).apply(T)
+DirichletBC(Q, 0, top_id).apply(T)
+# -
+
+# We now set up our output. To do so, we create the output file as a ParaView Data file
+# that uses the XML-based VTK file format. We also open a file for logging and
+# instantiate G-ADOPT geodynamical diagnostic utility.
+
+# +
+output_file = VTKFile("output.pvd")
+
+plog = ParameterLog("params.log", mesh)
+plog.log_str("step time dt u_rms entrainment")
+
+gd = GeodynamicalDiagnostics(z, T, bottom_id, top_id)
+material_area = material_interface_y * lx  # Quantity of tracked material
+entrainment_height = 0.2  # Height above which entrainment diagnostic is calculated
+# -
+
+# Here, we set up the variational problem for the Stokes, energy, and level-set
+# systems. The Stokes and energy systems depend on the approximation defined above,
+# and the level-set system includes both advection and reinitialisation.
+
+# +
+energy_solver = EnergySolver(
+    T, u, approximation, delta_t, ImplicitMidpoint, bcs=temp_bcs
+)
+
+stokes_solver = StokesSolver(
+    z,
+    T,
+    approximation,
+    bcs=stokes_bcs,
+    nullspace=Z_nullspace,
+    transpose_nullspace=Z_nullspace,
+)
+
+subcycles = 1  # Number of advection solves to perform within one time step
+level_set_solver = LevelSetSolver(psi, u, delta_t, eSSPRKs10p3, subcycles, epsilon)
+# Increase the reinitialisation time step to make up for the coarseness of the mesh
+level_set_solver.reini_params["tstep"] *= 20
+# -
+
+# Finally, we initiate the time loop, which runs until the simulation end time is
+# attained.
+
+# +
+step = 0  # A counter to keep track of looping
+output_counter = 0  # A counter to keep track of outputting
+time_end = 0.05
```

Comment:
Although the benchmark goes for this time, given we are under resolved, perhaps we only need to run the tutorial to 0.02 or similar (and we can flag that the true benchmark runs for longer with a comment)?